test/common_device_type.py

```diff
@@ -135,25 +167,47 @@ def instantiated_test(self, test=test):
 
                 @wraps(test)
                 def instantiated_test(self, test=test, dtype=dtype):
-                    return test(self, cls.device_type, dtype)
+                    device_arg = cls.default_device if not hasattr(test, 'devices_required') else cls.all_devices
+                    return test(self, device_arg(), dtype)
 
                 setattr(cls, dtype_test_name, instantiated_test)
 
 
 class CPUTestBase(DeviceTypeTestBase):
-    device_type = "cpu"
+    device_type = 'cpu'
 
 
 class CUDATestBase(DeviceTypeTestBase):
-    device_type = "cuda"
+    device_type = 'cuda'
     _do_cuda_memory_leak_check = True
     _do_cuda_non_default_stream = True
 
+    @classmethod
+    def default_device(cls):
+        return 'cuda:0'
+
+    @classmethod
+    def all_devices(cls):
+        default_device_num = int(cls.default_device().split(':')[1])
+        num_devices = torch.cuda.device_count()
+
+        devices = [cls.default_device()]
+        cuda_str = 'cuda:{0}'
+        non_default_devices = [cuda_str.format(num) for num in range(num_devices) if num != default_device_num]
+        devices.extend(non_default_devices)
+        return devices
+
     @classmethod
     def setUpClass(cls):
         # has_magma shows up after cuda is initialized
         torch.ones(1).cuda()
         cls.no_magma = not torch.cuda.has_magma
+        cls.current_device = torch.cuda.current_device()
+        torch.cuda.set_device(cls.default_device())
```

Comment:
hmm, this means I can't test with a different default device from 'cuda:0'?

Why don't we just invert this and use the current device for default device so we don't have to write it explicitly?

And we don't even need this chunk of code then I think, because you can figure out the current device in a device-type agnostic way by doing `torch.tensor([], device=device_type).device`

Reply:
I like using the current device as the default device for CUDA.

I like the idea of making the current device acquisition device type generic but I don't think it will apply to XLA so it would just be for CUDA, and CUDA needs to override setUpClass, anyway.

Reply:
So what does `torch.tensor([], device=device_type).device` give me for an xla tensor?

Reply:
'xla:0' which is actually your CPU

Reply:
is there some document that explains this design?